content/en-us/luau/type-checking.md

```diff
@@ -0,0 +1,352 @@
+
+```lua
+print(addLotsOfNumbers(1, 2, 3, 4, 5)) -- 15
+print(addLotsOfNumbers(1, 2, "3", 4, 5)) -- attempt to add string to number
+```
+
+Instead, it is possible to assign a type to the `...`, just like how other types are assigned.
+
+```lua
+local function addLotsOfNumbers(...: number)
+```
+
+And now, the second line raises a type error.
+
+```lua
+print(addLotsOfNumbers(1, 2, 3, 4, 5))
+print(addLotsOfNumbers(1, 2, "3", 4, 5)) -- TypeError: Type `string` could not be converted into `number`.
+```
+
+However, with a type definition, the following statement does not work.
+
+```lua
+type addLotsOfNumbers = (...: number) -> number -- Expected type, got ':'
+```
+
+Instead, use the syntax `...type` to define the variadic type.
+
+```lua
+type addLotsOfNumbers = (...number) -> number
+```
+
+## Unions and Intersections
+
+It is possible to define a type to be two or more types, using a union or intersection.
+
+```lua
+type NumberOrString = number | string
+type Object1 = {foo: string}
+type Object2 = {bar: number}
+type Object1and2 = Object1 & Object2 -- {foo: string} & {bar: number}
+```
+
+## Defining an Inferred Type
+
+The `typeof` function can be used in a type definition to define an inferred type.
+
+```lua
+type Car = typeof {
+	Speed = 0,
+	Wheels = 4
+} --> Car: {Speed: number, Wheels: number}
+```
+
+One way to use `typeof`, is to define a metatable type using `setmetatable` inside the `typeof` function.
+
+```lua
+type Vector = typeof(setmetatable({}::{
+	x: number,
+	y: number
+}, {}::{
+	__add: (Vector, Vector|number) -> Vector
+}))
+
+-- Vector + Vector would return a Vector type
+```
+
+## Generics
+
+Generics allow you to define at a most basic level, substitution types. Lets assume a basic object that creates a State object.
+
+```lua
+local State = {
+	Key = "TimesClicked",
+	Value = 0
+}
+```
+
+The type for this state object without generics would be.
+
+```lua
+type State = {
+	Key: string,
+	Value: number
+}
+```
+
+However, you may want `Value` to be a type based on the incoming value, this can be done with generics. Generic types use the following basic shape:
+
+```lua
+type GenericType<T> = T
+```
+
+The `<T>` denotes a type that can be set to anything. The best way to visualize this is as a substitution type. 
+
+```lua
+type List<T> = {T}
+
+local Names: List<string> = {"Bob", "Dan", "Mary"}
+local Fibbonacci: List<number> = {1, 1, 2, 3, 5, 8, 13}
+```
+
+Generics can also have multiple substitutions inside the brackets.
+
+```lua
+type Map<K, V> = {[K]: V}
+```
+
+Using the state object above, it can be reworked to use a generic type.
+
+```lua
+type State<T> = {
+	Key: string,
+	Value: T
+}
+```
+
+### Function Generics
+
+Functions can also use generics, however, they behave differently when inference is introduced into it, using the State example again, it can be designed to infer the value of `T`, from the function's incoming arguments.

```

Comment:
```suggestion
Functions can also use generics. The State example can infer the value of `T` from the function's incoming arguments.
```


> they behave differently when inference is introduced

Not entirely sure what is meant by this

Reply:
This could probably do an entire rephrase. The feature here is that the type of `T` is inferred from what's given to the function outside.

As the code example shows better, if I pass a `string` into a function such that has a type of `(T) -> State<T>`, the State will become `string`.

I'll suggest my own idea in a second